src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -5160,7 +5160,8 @@ public object VisitConvertExpression(ConvertExpressionAst convertExpressionAst)
                         new[] { temp },
                         BuildHashtable(hashTableAst.KeyValuePairs, temp, ordered: true));
                 }
-                if (typeName.FullName.Equals("PSCustomObject", StringComparison.OrdinalIgnoreCase))
+                if (typeName.FullName.Equals("PSCustomObject", StringComparison.OrdinalIgnoreCase)
+                    || typeName.FullName.Equals("System.Management.Automation.PSCustomObject", StringComparison.OrdinalIgnoreCase))
```

Comment:
Not sure about that: `System.Management.Automation.PSCustomObject` is a different type from `System.Management.Automation.PSObject` which we construct. Did you have a chance to figure out what's the difference between them?